pkg/controllers/workapplier/apply.go

```diff
@@ -0,0 +1,539 @@
+	if err != nil {
+		return false, fmt.Errorf("failed to marshal the manifest object into JSON: %w", err)
+	}
+	annotations[fleetv1beta1.LastAppliedConfigAnnotation] = string(lastAppliedManifestJSONBytes)
+	isLastAppliedAnnotationSet := true
+
+	if err := validation.ValidateAnnotationsSize(annotations); err != nil {
+		// If the annotation size exceeds the limit, Fleet will set the annotation to an empty string.
+		annotations[fleetv1beta1.LastAppliedConfigAnnotation] = ""
+		isLastAppliedAnnotationSet = false
+	}
+
+	manifestObj.SetAnnotations(annotations)
+	return isLastAppliedAnnotationSet, nil
+}
+
+// getFleetLastAppliedAnnotation returns the last applied annotation of a manifest object.
+func getFleetLastAppliedAnnotation(inMemberClusterObj *unstructured.Unstructured) []byte {
+	annotations := inMemberClusterObj.GetAnnotations()
+	if annotations == nil {
+		// The last applied annotation is not found in the live object; normally this should not
+		// happen, but Fleet can still handle this situation.
+		klog.Warningf("no annotations in the live object %s/%s", inMemberClusterObj.GroupVersionKind(), klog.KObj(inMemberClusterObj))
+		return nil
+	}
+
+	lastAppliedManifestJSONStr, found := annotations[fleetv1beta1.LastAppliedConfigAnnotation]
+	if !found {
+		// The last applied annotation is not found in the live object; normally this should not
+		// happen, but Fleet can still handle this situation.
+		klog.Warningf("the last applied annotation is not found in the live object %s/%s", inMemberClusterObj.GroupVersionKind(), klog.KObj(inMemberClusterObj))
+		return nil
+	}
+
+	return []byte(lastAppliedManifestJSONStr)
+}
+
+// setManifestHashAnnotation computes the hash of the provided manifest and sets an annotation of the
+// hash on the provided unstructured object.
+func setManifestHashAnnotation(manifestObj *unstructured.Unstructured) error {
+	cleanedManifestObj := discardFieldsIrrelevantInComparisonFrom(manifestObj)
+	manifestObjHash, err := resource.HashOf(cleanedManifestObj.Object)
+	if err != nil {
+		return err
+	}
+
+	annotations := manifestObj.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[fleetv1beta1.ManifestHashAnnotation] = manifestObjHash
+	manifestObj.SetAnnotations(annotations)
+	return nil
+}
+
+// setOwnerRef sets the expected owner reference (reference to an AppliedWork object)
+// on a manifest to be applied.
+func setOwnerRef(obj *unstructured.Unstructured, expectedAppliedWorkOwnerRef *metav1.OwnerReference) {
+	ownerRefs := obj.GetOwnerReferences()
+	if ownerRefs == nil {
+		ownerRefs = []metav1.OwnerReference{}
+	}
+	// Typically owner references is a system-managed field, and at this moment Fleet will
+	// clear owner references (if any) set in the manifest object. However, for consistency
+	// reasons, here Fleet will still assume that there might be some owner references set
+	// in the manifest object.
```

Comment:
TODO: try to preserve the original owner reference

Reply:
TO-DO added.